OpenRA.Mods.Common/Traits/World/SpawnMapActors.cs

```diff
@@ -38,11 +38,25 @@ public void WorldLoaded(World world, WorldRenderer wr)
 				var initDict = actorReference.InitDict;
 				initDict.Add(new SkipMakeAnimsInit());
 				initDict.Add(new SpawnedByMapInit(kv.Key));
+
+				if (PreventMapSpawn(world, actorReference))
+					continue;
+
 				var actor = world.CreateActor(actorReference.Type, initDict);
 				Actors[kv.Key] = actor;
 				LastMapActorID = actor.ActorID;
 			}
 		}
+
+		bool PreventMapSpawn(World world, ActorReference actorReference)
+		{
+			foreach (var pms in world.WorldActor.TraitsImplementing<IPreventMapSpawn>()
```

Comment:
A map may have several hundred actors, so it would be a good idea to do this lookup once and cache the results as an array before the ActorDefinitions loop rather than repeating for all actors.

Reply:
Ah, yes, you're right. That should have been kinda obvious, sorry. Updated.